AD Helpers: Introduce base class for cell-level AD helper classes

jppelteret opened the pull request:
Blocked by #6975

**TODO**:
- [x] Update documentation (headers)
- [x] Update documentation (source)
- [x] Check validity of discussions wrt. both `ADOL-C` and `Sacado`
- [x] Proofread

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] tbufsize ADOL-C Taylor buffer size
+       */
+      static void
+      enable_taping(const types::tape_index        tape_index,
```
I don't recall whether I commented on this already in the previous patch, but `enable/disable_` taping sounds like you make it possible. Is this what you mean? Or do you really mean that this *starts* or *stops* taping?

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+                     const bool              write_tapes_to_file);
+
+      /**
+       * Prints the statistics regarding the usage of the tapes.
```
-> Prints statistics?

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       */
+      static void
+      print_tape_stats(std::ostream &          stream,
+                       const types::tape_index tape_index);
```
I would revert the argument list to provide input before output arguments.

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+      /**
+       * Prints the statistics regarding the usage of the tapes.
+       *
+       * @param[in] stream The output stream to which the values are to be
```
It's really an output argument

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] gradient The scalar values of the dependent function's
+       *             gradients. It is expected that this vector be of the
```
I would just say "The values", not "scalar values"

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       *            dependent function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] hessian The scalar values of the dependent function's
```
same here

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] n_dependent_variables The number of dependent variables.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] values The scalar values of the dependent functions.
```
maybe "component" values, given that this is a vector-valued function that has what in deal.II we call "components"?

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       */
+      static void
+      values(const types::tape_index        active_tape_index,
+             const unsigned int             n_dependent_variables,
```
I'm surprised you have to specify the number of dependent variables. Doesn't the function being taped know that? Or Maybe I just don't know (yet) how all of this is going to be used...

jppelteret replied:
> Doesn't the function being taped know that?

Somehow not... This is the downside to passing pointers rather than objects. I could infer this information from the size of the `values` vector, but for safety I check that its been initialised to the right size.

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] n_dependent_variables The number of dependent variables.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] jacobian The scalar values of the dependent functions'
```
same with "scalar values". same in all of the places below where I'm not going to make it up any more

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * Computes the value of the scalar field.
+       *
+       * @param[in] dependent_variables The dependent variables whose values are
+       *            to be extracted.
```
I don't think I understand this argument. It's an input argument, so what is extracted from where? And why does one have to input a vector of dependent variables and get a single value out?

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] independent_variables The independent variables whose
+       *            sensitivities were tracked.
+       * @param[in] dependent_variables The (single) dependent variable whose
+       *            gradients are to be extracted.
```
same here: you say "single value", but it's actually an array

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * @param[in] independent_variables The independent variables whose
+       *            sensitivities were tracked.
+       * @param[in] dependent_variables The (single) dependent variable whose
+       *            Hessians are to be extracted.
```
and here

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+       * Computes the values of the vector field.
+       *
+       * @param[in] dependent_variables The dependent variables whose Hessians
+       *            are to be extracted.
```
Hessians -> values?

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+    // Note: In the case of ADOL-C taped numbers, the associated scalar
+    // type is always expected to be a double. So we need to make a further
+    // specialization when ScalarType is a float.
+    template <typename ADNumberType>
```
make this a doxygen-style comment so it shows up on the website.

you can copy the documentation of each of the functions using the `@copydoc` tag

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+    // Note: ADOL-C only has drivers for doubles, and so floats are
+    // not intrinsically supported. This wrapper struct works around
+    // the issue when necessary.
+    template <typename ADNumberType>
```
doxygen-style comment

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+      // A dummy function to define the active dependent variable when using
+      // reverse-mode AD.
+      template <typename ADNumberType>
+      static
```
It's not useful to mark global functions in `.h` files with `static`. same below

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+    {
+      // A dummy function to define the active dependent variable when using
+      // reverse-mode AD.
+      template <typename ADNumberType>
```
this function and the ones below are internal, but they should probably still be documented using doxygen-style comments

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+    // reverse mode to compute the first derivatives (and, if supported,
+    // forward mode for the second).
+    template <typename ADNumberType, typename ScalarType>
+    struct TapelessDrivers<
```
doxygen style for class docs, `@copydoc` for member docs in the same way as above

jppelteret replied:
This is all encapsulated in a `#ifndef DOXYGEN`, so we don't build documentation for the class specialisations.

jppelteret replied:
(That's [this line](https://github.com/dealii/dealii/pull/6976/files#diff-42fec071a85b1a667a206111406371e6R490))

bangerth commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+    // Specialization for auto-differentiable numbers that use
+    // forward mode to compute the first (and, if supported, second)
+    // derivatives.
+    template <typename ADNumberType, typename ScalarType>
```
same here

jppelteret commented:
@bangerth You've commented on stuff that's got nothing to do with this branch! This was in fact "reviewed", and I've yet to rebase this branch. 

Maybe what's best is if I attend to all of the remarks that you've made here and in #6975 before we continue. For my sanity...

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+     */
+    DeclExceptionMsg(
+      ExcRequiresAdolC,
+      "This function is only available if deal.II is compiled with ADOL-C.");
```
No, thats not correct. The `#ifdef DEAL_II_WITH_ADOLC` is closed [here](https://github.com/dealii/dealii/pull/6976/files/5a10a2342382a0c2637de05a3bca101b391aa05b#diff-42fec071a85b1a667a206111406371e6R41).

jppelteret commented on include/deal.II/differentiation/ad/ad_drivers.h:
```diff
@@ -0,0 +1,1583 @@
+      /**
+       * Typedef for tape buffer sizes.
+       */
+      using tape_buffer_sizes = unsigned int;
```
I'm going to leave this for another PR. It'll require some thought as to how we'd want to implement this.

jppelteret replied:
See #7117

jppelteret commented on cmake/modules/FindADOLC.cmake:
```diff
@@ -77,6 +77,18 @@ IF(EXISTS ${ADOLC_SETTINGS_H})
+  # Check whether ADOL-C is configured with tapeless number reference counting
+  #
+  FILE(STRINGS "${ADOLC_SETTINGS_H}" ADOLC_WITH_TAPELESS_REFCOUNTING_STRING
+    REGEX "#define USE_ADTL_REFCOUNTING 1"
```
See #7116

jppelteret commented:
@bangerth I've addressed your comments to the AD drivers header in #7118. May you please review them before anything else so that I can rebase #6975 on top of them? Thanks in advance!

jppelteret commented:
Although this is in part blocked by #6975, the `ADHelperCellLevelBase` class is ready for review.

jppelteret commented:
@bangerth Thanks for merging the other branch. This one is now ready for review.

jppelteret commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -314,6 +316,19 @@ namespace Differentiation
+       * defined that will have a sensitivity to the given independent
+       * variables. In the computation of $\mathbf{f}(\mathbf{X})$, this will
+       * be the number of outputs $\mathbf{f}$, i.e. the dimension of the
+       * domain or image space.
```
Right, the [Wikipedia article](https://en.wikipedia.org/wiki/Image_(mathematics)) helped clarify this (I think it must have been in high school that I last heard this terminology :-)

bangerth replied:
Took me a long time to remember this as well, but now that I'm teaching this stuff, I really have to know :-)

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that are used in,
```
are -> is

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * defined that will have a sensitivity to the given independent
+       * variables. In the computation of $\mathbf{f}(\mathbf{X})$, this will
+       * be the number of outputs $\mathbf{f}$, i.e. the dimension of the
+       * domain or image space.
```
move "domain" to the other variable's description

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * degree-of-freedom values on the current finite element. These define
+       * the values of all independent variables. When considering taped AD
+       * numbers with branching functions, to avoid potential issues with branch
+       * switching it may be a good idea to choose these values close to those
```
-> "close to or equal to" ?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       *
+       * @note The input value type must correspond to this class's number type.
+       * So if this class is templated on type double then the number type
+       * associated with value must also be of type double.
```
Hm, this class *is* templated. Do you mean to say "if the ... template argument of this class is @p double, then ..."?

Also, that's not how `scalar_type` is declared. It is not equal to the template argument of this class, but something more complicated.

jppelteret replied:
Indeed. I've now reformulated the message that I'm trying to convey here.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       *
+       * @param[in] values A global field from which the values of all
+       * independent variables will be extracted. This typically will be the
+       * solution vector around at which point a residual vector is to be
```
drop "at"

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * computed and around which linearization is to occur.
+       * When considering taped AD numbers with branching functions, to avoid
+       * potential issues with branch switching it may be a good idea to choose
+       * these values close to those that will be later evaluated and linearized
```
"close to or equal to"

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * these values close to those that will be later evaluated and linearized
+       * around.
+       * @param[in] local_dof_indices A vector of degree-of-freedom indices from
+       * which to extract the local degree-of-freedom values.
```
maybe also say here that this would then be what `cell->get_dof_indices()` would typically return, since you also give an example for the other argument

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * @note For taped AD numbers, this operation is only valid in recording mode.
+       */
+      const std::vector<ad_type> &
+      get_sensitive_dof_values();
```
maybe say in the documentation that this is the set of variables with which one would then perform computations, and based on which one can then extract both the value of the function and its derivatives with the member functions below?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * The values of the components of the returned object are initialized to
+       * the values set with register_dof_values().
+       *
+       * @return An object of auto-differentiable type numbers representing the
```
object -> array?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       *
+       * @note The input value type must correspond to this class's number type.
+       * So if this class is templated on type double then the number type
+       * associated with value must also be of type double.
```
same comment as above

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
```
maybe say "a separate call"?

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
```
same here

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+
+      /**
+       * Computes the value of the residual vector field
+       * $\mathbf{r}(\mathbf{X})$.
```
I guess this function does not actually compute anything, but it should do so in derived classes?

It would probably be useful to state how large the vector is supposed to be that's returned. Also, we generally return `Vector` objects through arguments, because they're reused from quadrature point to quadrature point.

Same for the next function.

jppelteret replied:
> we generally return Vector objects through arguments

Well that change going to break a *few* tests...

jppelteret replied:
Actually, I'm not happy with making this change with the current class interface. 

The problem is that the size of the returned objects (for `ADHelperVariationalFormulation` and `ADHelperResidualLinearisation`) differ. For the former, the residual vector has length `n_independent_variables` (because its the first derivative of a scalar function, the only dependent variable) and the latter has length `n_dependent_variables` . 

The only way I see this working (i.e. having the output passed as an argument) is if we have another set of virtual functions that return the expected sizes of these objects. If you like this idea then do you have a suggestion for a name (I think something like `m()` and `n()` is not descriptive, and maybe `n_rows_residual`, `n_rows_linearization` and `n_cols_linearization` is too verbose.

bangerth replied:
If you feel strongly, then let's leave it as is.

The argument for returning via an argument is that it doesn't require memory allocation for every call, whereas when you return by value, you'll have to allocate an object. I can't say for sure whether that's a noticeable overhead, but we've generally followed this strategy.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       * values.
+       */
+      virtual Vector<scalar_type>
+      compute_residual() const = 0;
```
I think I understand this class a lot better now that I see these two member functions. But to understand their names and purpose, I think it would be useful to show a common use case in the class documentation -- namely, how one would implement a derived class. Do you happen to have an example?

jppelteret replied:
I will be furnishing each one of the most derived helper classes with a full example (see [here in the next PR](https://github.com/dealii/dealii/pull/6977/files#diff-f54538107ff9ecaf63f8d20799fa3ca7R1011), for example). In total you may expect 5 code samples (the snippet in the very base class and one detailed example for each of the four helper classes). Do you still want me to add another example here? Its seems like a bit of overkill to me.

bangerth replied:
Yes, that would be overkill. Just leave a note in the class documentation that derived classes will have code examples that show how to use this interface -- that ought to be enough.

bangerth commented on include/deal.II/differentiation/ad/ad_helpers.h:
```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+#    ifdef DEBUG
+      for (unsigned int i = 0; i < this->n_independent_variables(); ++i)
+        {
+          Assert(this->registered_independent_variable_values[i] == false,
```
Just remove the `#ifdef` -- the compiler will optimize away the loop once the `Assert` is expanded to nothing in release mode.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -725,6 +725,120 @@ namespace Differentiation
+    template <enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
+    ADHelperCellLevelBase<ADNumberTypeCode,
+                          ScalarType>::~ADHelperCellLevelBase()
+    {}
```
you could just mark this destructor as `=default` in the class declaration.

bangerth commented on source/differentiation/ad/ad_helpers.cc:
```diff
@@ -725,6 +725,120 @@ namespace Differentiation
+#  ifdef DEBUG
+      for (unsigned int i = 0; i < this->n_independent_variables(); ++i)
+        {
+          Assert(this->registered_independent_variable_values[i] == false,
```
same as in the .h file. possibly also in other places.

jppelteret commented:
Thanks for your comments @bangerth! I've addressed all of them except for two in my local copy. Once we have a resolution on https://github.com/dealii/dealii/pull/6976#discussion_r214192962 and https://github.com/dealii/dealii/pull/6976#discussion_r214470933 then I'll make the final changes and rebase this branch.

bangerth commented:
I think we have a resolution for the two issues in question -- in particular, you can leave the return-vector-by-value as is right now. We can always change this later if we feel strongly, but for the moment I think moving forward is what's more relevant :-)

jppelteret commented:
@bangerth Thanks for checking up on this. I'd been thinking more about https://github.com/dealii/dealii/pull/6976#discussion_r214772814 and I now think that I can make this change, but I'll just need to make clear in the documentation of each derived class what the object sizes will be. I've yet to find a free moment to make the changes here and to my tests. I'll try to get to them on the weekend and will let you know when its done.

bangerth commented:
OK. I'm not trying to push, just inquiring :-) Let me know if/once you want me to take another look!

jppelteret commented:
@bangerth I found some time this morning to implement the rest of the changes. I think that this is ready for another look. Thanks in advance!

jppelteret commented:
/run-tests

bangerth commented:
Great, thanks! 

Rebase the next PR then and let me know once you want me to take a look at it!
